### PR TITLE
Port `cumsum` kernel to structured kernels.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61899 Port `cumprod` kernel to structured kernels.
* **#61842 Port `cumsum` kernel to structured kernels.**

Tracking issue: #55070

This PR also removes `at::_cumsum`, which was the "backend" for `at::cumsum`.

Differential Revision: [D29939152](https://our.internmc.facebook.com/intern/diff/D29939152)

### CODE REVIEW

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29939152).

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -116,6 +116,31 @@ TORCH_META_FUNC(argmin)
...
+  // Checking whether 'dim' is valid.
+  maybe_wrap_dim(dim, self.dim());
+
+  const auto& result = maybe_get_output();
+  ScalarType out_dtype;
+
+  if (result.defined()) {
+    out_dtype = dtype.value_or(result.scalar_type());
+    // This check is still here because the inline version of structured kernels
+    // does not do any checks on 'set_output'.
```
Comment (ezyang) [inline aten/src/ATen/native/ReduceOps.cpp:130]: I'm not sure I understand. We do test that the passed dtype matches the out dtype; sample generated code:

```
    void set_output(int64_t output_idx, IntArrayRef sizes, IntArrayRef strides,
                    TensorOptions options, DimnameList names) override {
        
        const auto& out = outputs_[output_idx].get();
        TORCH_CHECK(options.dtype() == out.dtype(),
            "Expected out tensor to have dtype ", options.dtype(), ", but got ", out.dtype(), " instead");        TORCH_CHECK(options.device() == out.device(),
            "Expected out tensor to have device ", options.device(), ", but got ", out.device(), " instead");     
        bool resized = at::native::resize_output(outputs_[output_idx], sizes);

```

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -180,53 +205,21 @@ Tensor& logcumsumexp_out(const Tensor& self, int64_t dim, Tensor& result) {
...
+ const Tensor& result) {
+  NoNamesGuard guard;
+  if (self.dim() == 0) {
+    result.fill_(self);
+  } else if (self.numel() == 0) {
+    result.zero_();
+  } else {
+    dim = maybe_wrap_dim(dim, self.dim());
+    auto& result_mut = const_cast<Tensor&>(result);
+    cumsum_stub(self.device().type(), result_mut, self.to(result.scalar_type()), dim);
```
Comment (ezyang) [inline aten/src/ATen/native/ReduceOps.cpp:221]: If possible, changing the type signature of `cumsum_stub` would be better, as these tend to be fairly localized changes.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29939152).

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29939152).

File: aten/src/ATen/native/ReduceOps.cpp (hunk 1 above)
Comment (ysiraichi) [reply aten/src/ATen/native/ReduceOps.cpp:130]: Ah. Replace `inline` => `inplace`. In the inplace version, we don't test it, do we?

Comment (ezyang) [reply aten/src/ATen/native/ReduceOps.cpp:130]: Oh yeah, inplace still not testing; see also https://github.com/pytorch/pytorch/issues/57827

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29939152).

**Comment by ysiraichi** [comment]
The 2 CI failures do not seem to be related to this PR (not sure about the Facebook internal tests, though).

**Pull request closed without merging**